[ADR] Use raw data in commands and events (#13787)
Sylius/Sylius: adr/2022_03_18_raw_data_in_command_and_events.md

# Use raw data in commands and events

* Status: accepted
* Date: 2022-03-18

## Context and Problem Statement

> While defining new commands or events we may define them with entities or any objects. Objects passed to commands or

Reviewer: This is more of my semantical taste, but this section seems to have mixed context and description of the proposed solution. I'm fine with it as it is, but what could we consider for this (or future) ADRs is to focus in this section on the problem itself, like:
```
While working with commands and handlers, we are always working with objects; therefore, these objects need to be present in handlers. We may either fetch them from the repository based on identifiers or pass them as a part of commands. The selected solution needs to take into account that our commands may be dispatched to external systems or possessed asynchronously
```

and then split the current content that is written here into a description of each possible choice

events may desynchronize with the main object which can be causing problems with handlers that may operate on wrong data.
Passing objects to commands can ease us to implement new features and make it faster but this solution gives us plenty of
problems with integrations with external systems or eg. operating with the actual state of an object especially when
we are doing it asynchronously.

## Decision Drivers

* always operate on the actual state of the object
* ease to integrate with external systems

> * to be more consistent with CQRS pattern

Reviewer: ```suggestion
* be more consistent with CQRS pattern
```

## Considered Options

### Passing raw data or objects to commands

> * good, because it can be ease for us to implement new features

Reviewer: ```suggestion
* good, because it can be easier for us to implement new features
```

> * bad, because it can be problematicaly when we are using async communication

Reviewer 2: ```suggestion
* bad, because it can be problematic when we are using async communication
```

> * bad, because it can be hard to integrate commands written like that with extenal systems

Reviewer 2: ```suggestion
* bad, because it can be hard to integrate commands written like that with external systems
```

### Passing only raw data or identifiers of the object(s) we want to change etc.

* good, because there is no problem with desynchronizing data while async communication

> * good, because we can easy integrate with extenal systems

Reviewer 2: ```suggestion
* good, because we can easily integrate with external systems
```

# Use raw data in commands and events

* Status: accepted
* Date: 2022-03-18

## Context and Problem Statement

While working with commands and handlers, we are always working with objects; therefore, these objects need to be 
present in handlers. We may either fetch them from the repository based on identifiers or pass them as a part of 
commands. The selected solution needs to take into account that our commands may be dispatched to external systems or 

> possessed asynchronously

Reviewer: 👻 
```suggestion
processed asynchronously
```

Passing objects to commands can ease us to implement new features and make it faster but this solution gives us plenty of
problems with integrations with external systems or eg. operating with the actual state of an object especially when
we are doing it asynchronously.

## Decision Drivers

* always operate on the actual state of the object
* ease to integrate with external systems
* to be more consistent with CQRS pattern

## Considered Options

### Passing raw data or objects to commands

* good, because it can be ease for us to implement new features
* bad, because it can be problematicaly when we are using async communication
* bad, because it can be hard to integrate commands written like that with extenal systems

### Passing only raw data or ids of object we want to change etc.

* good, because there is no problem with desynchronizing data while async communication
* good, because we can easy integrate with extenal systems
* good, because it's more consistent with CQRS pattern
* bad, because we need to always make queries to DB to get objects we want to modify (but we are caching queries from 
DB when we are using doctrine as our ORM)

## Decision outcome

> Chosen option: **"Passing only raw data or ids of object we want to change etc."**

Reviewer 3: ```suggestion
Chosen option: **"Passing only raw data or identifiers of the object(s) we want to change etc."**
```
